module/support/optimisation/NSGA2Evaluator/src/NSGA2Evaluator.cpp

```diff
@@ -0,0 +1,318 @@
+            case State::Value::SETTING_UP_TRIAL: return transition_events(event);
+            case State::Value::RESETTING_TRIAL: return transition_events(event);
+            case State::Value::EVALUATING: return transition_events(event);
+            case State::Value::TERMINATING_EARLY: return transition_events(event);
+            case State::Value::TERMINATING_GRACEFULLY: return transition_events(event);
+            case State::Value::FINISHED:
+                // Arguably this should return FINISHED regardless of event, unless we want to be able to
+                // reset
+                if (event.value == Event::Value::FITNESS_SCORES_SENT) {
+                    return State(State::Value::FINISHED);
+                }
+                else {
+                    return transition_events(event);
+                }
+
+            default: return State(State::Value::UNKNOWN);
+        }
+    }
+
+    NSGA2Evaluator::State NSGA2Evaluator::transition_events(NSGA2Evaluator::Event event) {
+        switch (event.value) {
+            case Event::Value::EVALUATE_REQUEST: return State(State::Value::SETTING_UP_TRIAL);
+            case Event::Value::CHECK_READY: return State(State::Value::WAITING_FOR_REQUEST);
+            case Event::Value::TERMINATE_EVALUATION: return State(State::Value::FINISHED);
+            case Event::Value::TRIAL_SETUP_DONE: return State(State::Value::RESETTING_TRIAL);
+            case Event::Value::RESET_DONE: return State(State::Value::EVALUATING);
+            case Event::Value::TERMINATE_EARLY: return State(State::Value::TERMINATING_EARLY);
+            case Event::Value::TRIAL_COMPLETED: return State(State::Value::TERMINATING_GRACEFULLY);
+            case Event::Value::FITNESS_SCORES_SENT: return State(State::Value::WAITING_FOR_REQUEST);
+
+            default: return State(State::Value::UNKNOWN);
+        }
+    }
+
+    void NSGA2Evaluator::walk(Eigen::Vector3d vec3) {
+        emit<Task>(std::make_unique<Walk>(vec3), 1);
+    }
+
+    // Handle the WAITING_FOR_REQUEST state
+    void NSGA2Evaluator::waiting_for_request() {
+        log<NUClear::DEBUG>("Waiting For Request");
+        emit(std::make_unique<NSGA2EvaluatorReady>());  // Let the optimiser know we're ready
+    }
+
+    // Handle the SETTING_UP_TRIAL state
+    void NSGA2Evaluator::setting_up_trial() {
+        log<NUClear::DEBUG>("Setting Up Trial");
+
+        generation = last_eval_request_msg.generation;
+        individual = last_eval_request_msg.id;
+
+        if (last_eval_request_msg.task == "walk") {
+            task = std::make_unique<WalkEvaluator>();
+        }
+        else if (last_eval_request_msg.task == "strafe") {
+            task = std::make_unique<StrafeEvaluator>();
+        }
+        else if (last_eval_request_msg.task == "rotation") {
+            task = std::make_unique<RotationEvaluator>();
+        }
+        else {
+            log<NUClear::ERROR>("Unhandled task type:", last_eval_request_msg.task);
+        }
```

Comment:
while using an enum makes it a little better I think this whole thing needs generalising so you don't have to keep adding code in so many places every time you make a new problem for the optimiser. also going through this I dont really get why there are all these different tasks since there should be only one task with a configurable walk command?